crates/fluvio-hub-util/src/fvm/api/client.rs

```diff
@@ -0,0 +1,190 @@
+    pub async fn fetch_package_set(
+        &self,
+        name: impl AsRef<str>,
+        channel: &Channel,
+        arch: &str,
+    ) -> Result<PackageSet> {
+        let url = self.make_fetch_package_set_url(name, channel, arch)?;
+        let mut res = get(url).await.map_err(|err| Error::msg(err.to_string()))?;
+        let pkg = res.body_json::<PackageSet>().await.map_err(|err| {
+            Error::msg(format!(
+                "Server responded with status code {}",
+                err.status()
+            ))
+        })?;
+
+        Ok(pkg)
+    }
+
+    /// Downloads binaries from the Hub into the specified `target_dir`.
+    /// Returns the [`PackageSet`] that was downloaded.
+    pub async fn download_package_set(
+        &self,
+        name: impl AsRef<str>,
+        channel: &Channel,
+        arch: &str,
+        target_dir: PathBuf,
+    ) -> Result<PackageSet> {
+        let pkgset = self.fetch_package_set(name, channel, arch).await?;
+
+        for artf in pkgset.artifacts.iter() {
+            let mut res = surf::get(&artf.download_url)
+                .await
+                .map_err(|err| Error::msg(err.to_string()))?;
+
+            if res.status() == StatusCode::Ok {
+                let out_path = target_dir.join(&artf.name);
+                let mut file = File::create(&out_path)?;
+                let mut buf = Cursor::new(
+                    res.body_bytes()
+                        .await
+                        .map_err(|err| Error::msg(err.to_string()))?,
+                );
+
+                copy(&mut buf, &mut file)?;
+                self.checksum_artifact(artf, &file).await?;
+
+                tracing::debug!(
+                    "Artifact downloaded: {} at {:?}",
+                    artf.name,
+                    out_path.display()
+                );
+
+                continue;
+            }
+
+            tracing::warn!(
+                "Failed to download artifact {}@{} from {}",
+                artf.name,
+                artf.version,
+                artf.download_url
+            );
+        }
+
+        Ok(pkgset)
+    }
+
+    /// Verifies downloaded artifacts checksums against the upstream checksums
+    async fn checksum_artifact(&self, artf: &Artifact, local: &File) -> Result<()> {
+        let local_file_shasum = Self::shasum256(local)?;
+        let upstream_shasum = surf::get(&artf.sha256_url)
+            .await
+            .map_err(|err| Error::msg(err.to_string()))?
+            .body_string()
+            .await
+            .map_err(|err| Error::msg(err.to_string()))?;
+
+        if local_file_shasum != upstream_shasum {
+            return Err(Error::msg(format!(
+                "Artifact {} didnt matched upstream shasum. {} != {}",
+                artf.name, local_file_shasum, upstream_shasum
+            )));
+        }
+
+        Ok(())
+    }
+
+    /// Generates the Sha256 checksum for the specified file
+    fn shasum256(file: &File) -> Result<String> {
```

Comment:
what's differ betwen this and other checksum?

Reply:
This function is no more, I centralized in `utils` as `sha256_digest` which generates Sha256 checksum for a given file.

And now it receives the path to the file.